File: frontend/src/services/denuncias.ts
const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export type CrearDenunciaPayload = {
  aviso_id: string;
  motivo: string;
  comentario?: string;
};

export type AvisoDenunciado = {
  id: string;
  titulo?: string;
  descripcion?: string;
  telefono?: string;
  whatsapp?: string;
  zona?: string;
  distrito?: string;
  estado?: string;
  activo?: boolean;
  fecha_expiracion?: string;
  motivo_baja?: string;
};

export type Denuncia = {
  id: string;
  aviso_id: string;
  motivo: string;
  comentario?: string | null;
  estado: string;
  accion_admin?: string | null;
  revisada_at?: string | null;
  created_at: string;
  aviso?: AvisoDenunciado | null;
};

export async function crearDenuncia(payload: CrearDenunciaPayload) {
  const response = await fetch(`${API_URL}/api/denuncias/`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    throw new Error(data?.detail || "No se pudo enviar la denuncia");
  }

  return data;
}

export async function listarDenuncias(
  adminToken: string,
  estado: string = "pendiente"
): Promise<Denuncia[]> {
  const response = await fetch(
    `${API_URL}/api/denuncias/?estado=${encodeURIComponent(estado)}`,
    {
      headers: {
        "X-Admin-Token": adminToken,
      },
      cache: "no-store",
    }
  );

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    throw new Error(data?.detail || "No se pudieron cargar las denuncias");
  }

  return data;
}

export async function marcarDenunciaRevisada(
  denunciaId: string,
  adminToken: string
) {
  const response = await fetch(
    `${API_URL}/api/denuncias/${denunciaId}/revisar`,
    {
      method: "PATCH",
      headers: {
        "X-Admin-Token": adminToken,
      },
    }
  );

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    throw new Error(data?.detail || "No se pudo marcar como revisada");
  }

  return data;
}

export async function darBajaAvisoDesdeDenuncia(
  denunciaId: string,
  motivoBaja: string,
  adminToken: string
) {
  const response = await fetch(
    `${API_URL}/api/denuncias/${denunciaId}/dar-baja-aviso`,
    {
      method: "PATCH",
      headers: {
        "Content-Type": "application/json",
        "X-Admin-Token": adminToken,
      },
      body: JSON.stringify({
        motivo_baja: motivoBaja,
      }),
    }
  );

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    throw new Error(data?.detail || "No se pudo dar de baja el aviso");
  }

  return data;
}

File: backend/app/routers/denuncias.py
from fastapi import APIRouter, Header, HTTPException, Query
from pydantic import BaseModel

from app.core.config import ADMIN_IMPORT_TOKEN
from app.services.denuncia_service import (
    crear_denuncia,
    dar_baja_aviso_por_denuncia,
    listar_denuncias,
    marcar_denuncia_revisada,
)


router = APIRouter()


class DenunciaCreate(BaseModel):
    aviso_id: str
    motivo: str
    comentario: str | None = None


class BajaAvisoDenunciaRequest(BaseModel):
    motivo_baja: str


def validar_admin_token(x_admin_token: str | None):
    if not ADMIN_IMPORT_TOKEN:
        raise HTTPException(
            status_code=500,
            detail="ADMIN_IMPORT_TOKEN no está configurado"
        )

    if not x_admin_token or x_admin_token != ADMIN_IMPORT_TOKEN:
        raise HTTPException(
            status_code=401,
            detail="No autorizado"
        )


@router.post("/")
def denunciar_aviso(denuncia: DenunciaCreate):
    try:
        return crear_denuncia(denuncia.model_dump())

    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error))

    except Exception as error:
        raise HTTPException(status_code=500, detail=str(error))


@router.get("/")
def obtener_denuncias(
    estado: str | None = Query(default="pendiente"),
    x_admin_token: str | None = Header(default=None)
):
    validar_admin_token(x_admin_token)

    try:
        return listar_denuncias(estado)

    except Exception as error:
        raise HTTPException(status_code=500, detail=str(error))


@router.patch("/{id}/revisar")
def revisar_denuncia(
    id: str,
    x_admin_token: str | None = Header(default=None)
):
    validar_admin_token(x_admin_token)

    try:
        return marcar_denuncia_revisada(id)

    except Exception as error:
        raise HTTPException(status_code=500, detail=str(error))


@router.patch("/{id}/dar-baja-aviso")
def baja_aviso_denunciado(
    id: str,
    data: BajaAvisoDenunciaRequest,
    x_admin_token: str | None = Header(default=None)
):
    validar_admin_token(x_admin_token)

    try:
        return dar_baja_aviso_por_denuncia(
            denuncia_id=id,
            motivo_baja=data.motivo_baja,
            bajado_por="admin"
        )

    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error))

    except Exception as error:
        raise HTTPException(status_code=500, detail=str(error))